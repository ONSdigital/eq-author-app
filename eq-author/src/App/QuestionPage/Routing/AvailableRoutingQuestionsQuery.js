import React from "react";
import PropTypes from "prop-types";
import { Query } from "react-apollo";
import gql from "graphql-tag";

const GET_AVAILABLE_ROUTING_QUESTIONS = gql`
  query GetAvailableRoutingQuestions($id: ID!) {
    questionPage(id: $id) {
      id
      displayName
      availableRoutingQuestions {
        id
        displayName
        section {
          id
          displayName
        }
      }
    }
  }
`;

const AvailableRoutingQuestionsQuery = ({ pageId, children }) => (
  <Query
    query={GET_AVAILABLE_ROUTING_QUESTIONS}
    variables={{ id: pageId }}
    fetchPolicy="cache-and-network"
  >
    {children}
  </Query>
);

AvailableRoutingQuestionsQuery.propTypes = {
  pageId: PropTypes.string.isRequired,
  children: PropTypes.func.isRequired
};

export default AvailableRoutingQuestionsQuery;
