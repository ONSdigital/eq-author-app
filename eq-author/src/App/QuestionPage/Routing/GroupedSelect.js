import React from "react";
import PropTypes from "prop-types";
import { isNil } from "lodash";
import Select from "components/Forms/Select";

export const textSelect = "Please selectâ€¦";

const GroupedSelect = ({ groups, onChange, value, valid, ...otherProps }) => {
  const optGroups = groups.map(group => (
    <optgroup label={group.label} key={group.id}>
      {group.options.map(option => (
        <option
          value={option.value}
          key={option.value}
          disabled={option.disabled}
        >
          {option.label}
        </option>
      ))}
    </optgroup>
  ));

  return (
    <Select
      onChange={onChange}
      value={isNil(value) ? "" : value}
      invalid={!valid}
      {...otherProps}
    >
      {isNil(value) && (
        <option disabled value="" key="please-select">
          {textSelect}
        </option>
      )}
      {optGroups}
    </Select>
  );
};

const OptionProp = PropTypes.shape({
  label: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired
});

const GroupProp = PropTypes.shape({
  label: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  options: PropTypes.arrayOf(OptionProp).isRequired
});

GroupedSelect.propTypes = {
  groups: PropTypes.arrayOf(GroupProp).isRequired,
  onChange: PropTypes.func.isRequired,
  value: PropTypes.string,
  valid: PropTypes.bool
};

GroupedSelect.defaultProps = {
  valid: true
};

export default GroupedSelect;
