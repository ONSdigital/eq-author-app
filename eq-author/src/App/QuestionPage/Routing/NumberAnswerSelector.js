import React from "react";
import styled from "styled-components";
import { colors, radius } from "constants/theme";
import { Number, Select, Label } from "components/Forms";
import PropTypes from "prop-types";
import { CURRENCY } from "constants/answer-types";
import CustomPropTypes from "custom-prop-types";
import VisuallyHidden from "components/VisuallyHidden";

export const RoutingNumberInput = styled(Number)`
  border-radius: ${radius};
  width: 8em;
  outline: none;
  ${props =>
    props.answerType === CURRENCY &&
    `
    padding-left: 1.2em;
  `};
`;

export const ComparatorSelector = styled(Select)`
  width: auto;
`;

const Value = styled.div`
  flex: 1 1 auto;
  display: flex;
  margin-left: 1em;
  align-items: center;
  position: relative;
`;

const CurrencySymbol = styled.div`
  position: absolute;
  opacity: 0.5;
  left: 0.5em;
`;

const NumberAnswerRoutingSelector = styled.div`
  display: flex;
  align-items: center;
  border: 1px solid ${colors.lightGrey};
  margin: 1em 0;
  border-radius: ${radius};
  padding: 1em;
`;

const NumberAnswerValueSelector = ({
  condition,
  onComparatorChange,
  handleValueChange
}) => {
  const otherProps = {
    questionPageId: condition.questionPage.id
  };

  return (
    <NumberAnswerRoutingSelector>
      <VisuallyHidden>
        <Label
          htmlFor={`routing-condition-comparator-${condition.routingValue.id}`}
        >
          Operator
        </Label>
      </VisuallyHidden>
      <ComparatorSelector
        id={`routing-condition-comparator-${condition.routingValue.id}`}
        onChange={value => onComparatorChange(otherProps, value)}
        name="comparator-select"
        value={condition.comparator}
        data-test="comparator-selector"
      >
        <option value="Equal">(=) Equal to</option>
        <option value="NotEqual">(&ne;) Not equal to</option>
        <option value="GreaterThan">(&gt;) More than</option>
        <option value="LessThan">(&lt;) Less than</option>
        <option value="GreaterOrEqual">(&ge;) More than or equal to</option>
        <option value="LessOrEqual">(&le;) Less than or equal to</option>
      </ComparatorSelector>
      <Value>
        <VisuallyHidden>
          <Label
            htmlFor={`routing-condition-value-${condition.routingValue.id}`}
          >
            Value
          </Label>
        </VisuallyHidden>
        {condition.answer.type === CURRENCY && (
          <CurrencySymbol>Â£</CurrencySymbol>
        )}
        <RoutingNumberInput
          id={`routing-condition-value-${condition.routingValue.id}`}
          min={-99999999}
          max={999999999}
          placeholder="Value"
          value={condition.routingValue.numberValue}
          name={condition.routingValue.id}
          onChange={handleValueChange}
          data-test="number-value-input"
          answerType={condition.answer.type}
        />
      </Value>
    </NumberAnswerRoutingSelector>
  );
};

NumberAnswerValueSelector.propTypes = {
  condition: PropTypes.shape({
    id: PropTypes.string.isRequired,
    comparator: PropTypes.string.isRequired,
    questionPage: PropTypes.shape({
      id: PropTypes.string.isRequired,
      displayName: PropTypes.string
    }),
    answer: CustomPropTypes.answer,
    routingValue: PropTypes.shape({
      id: PropTypes.string,
      value: PropTypes.array,
      numberValue: PropTypes.number
    })
  }).isRequired,
  onComparatorChange: PropTypes.func.isRequired,
  handleValueChange: PropTypes.func.isRequired
};

export default NumberAnswerValueSelector;
