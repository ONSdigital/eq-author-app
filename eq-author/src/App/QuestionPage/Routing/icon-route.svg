<?xml version="1.0" encoding="UTF-8"?>
<svg width="32px" height="32px" viewBox="0 0 32 32" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
    <!-- Generator: Sketch 50.2 (55047) - http://www.bohemiancoding.com/sketch -->
    <title>icon-route</title>
    <desc>Created with Sketch.</desc>
    <defs></defs>
    <g id="icon-route" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd">
        <polygon id="call_split---material" fill="#4A4A4A" fill-rule="nonzero" points="17.6504483 23.6277477 19.8429574 21.4353412 14.7956188 16.4110762 14.7956188 8.37225231 16.668387 8.37225231 16.668387 15.5889238 21.1675983 20.1107622 23.3601074 17.9183557 23.3601074 23.6277477"></polygon>
        <polygon id="Path" fill="#4A4A4A" fill-rule="nonzero" points="14.2360073 23.6277477 8.63989258 23.6277477 8.63989258 18.031633 10.7888006 20.180541 13.4973202 17.4720215 14.7956188 18.7703201 12.0870993 21.4788396"></polygon>
    </g>
</svg>