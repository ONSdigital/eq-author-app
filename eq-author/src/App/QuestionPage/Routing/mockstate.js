export default [
  {
    title: "1. Integer posuere erat a ante",
    displayName: "1. Integer posuere erat a ante",
    pages: [
      {
        title: "1.1 Curabitur blandit tempus porttitor",
        displayName: "1.1 Curabitur blandit tempus porttitor",
        pageType: "QuestionPage",
        options: [
          { label: "Bibendum Cras", id: "3", selected: false },
          { label: "Aenean Quam Venenatis Elit", id: "4", selected: false },
          { label: "Purus Ornare Ipsum", id: "5", selected: false },
          { label: "Nibh", id: "6", selected: false },
          { label: "Risus Mattis Magna Bibendum", id: "7", selected: false },
          {
            label: "Consectetur Ultricies Purus Elit Fermentum",
            id: "8",
            selected: false
          }
        ],
        id: "2",
        disabled: false
      },
      {
        title: "1.2 Nulla vitae elit libero, a pharetra augue.",
        displayName: "1.2 Nulla vitae elit libero, a pharetra augue.",
        pageType: "QuestionPage",
        options: [
          { label: "Yes", id: "10", selected: false },
          { label: "No", id: "11", selected: false }
        ],
        id: "9",
        disabled: false
      }
    ],
    id: "1"
  }
];
