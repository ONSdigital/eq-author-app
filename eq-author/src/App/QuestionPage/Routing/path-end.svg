<?xml version="1.0" encoding="UTF-8"?>
<svg width="43px" height="1000px" viewBox="0 0 43 1000" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
	<g id="Artboard" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd" stroke-linecap="square">
		<path d="M21,0 L21,500 M21,500 L21,500 M43,500 L22,500" id="Shape" stroke="#c3c3c3">
		</path>
	</g>
</svg>
