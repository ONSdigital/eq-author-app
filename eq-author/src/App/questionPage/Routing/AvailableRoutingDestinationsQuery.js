import React from "react";
import PropTypes from "prop-types";
import { Query } from "react-apollo";
import gql from "graphql-tag";

import LogicalDestination from "graphql/fragments/logical-destination.graphql";
import QuestionPageDestination from "graphql/fragments/question-page-destination.graphql";
import SectionDestination from "graphql/fragments/section-destination.graphql";

const GET_AVAILABLE_ROUTING_DESTINATIONS = gql`
  query GetAvailableRoutingDestinations($input: QueryInput!) {
    questionPage(input: $input) {
      id
      availableRoutingDestinations {
        logicalDestinations {
          ...LogicalDestination
        }
        questionPages {
          ...QuestionPageDestination
        }
        sections {
          ...SectionDestination
        }
      }
    }
  }

  ${LogicalDestination}
  ${QuestionPageDestination}
  ${SectionDestination}
`;

const AvailableRoutingDestinationsQuery = ({ pageId, children }) => (
  <Query
    query={GET_AVAILABLE_ROUTING_DESTINATIONS}
    variables={{ input: { pageId } }}
    fetchPolicy="cache-and-network"
  >
    {children}
  </Query>
);

AvailableRoutingDestinationsQuery.propTypes = {
  pageId: PropTypes.string.isRequired,
  children: PropTypes.func.isRequired,
};

export default AvailableRoutingDestinationsQuery;
