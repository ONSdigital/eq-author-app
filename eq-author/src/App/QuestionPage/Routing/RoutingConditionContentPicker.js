import React from "react";
import PropTypes from "prop-types";
import { get } from "lodash";

import ContentPickerSelect from "App/components/ContentPickerSelect";
import { QUESTION } from "App/components/ContentPickerSelect/content-types";
import { shapePageTree } from "App/components/ContentPicker/shapeTree";

import AvailableRoutingQuestionsQuery from "App/QuestionPage/Routing/AvailableRoutingQuestionsQuery";

export const UnwrappedRoutingConditionContentPicker = ({
  data,
  path,
  ...otherProps
}) => (
  <ContentPickerSelect
    name="questionPageId"
    contentTypes={[QUESTION]}
    questionData={shapePageTree(get(data, path))}
    {...otherProps}
  />
);

UnwrappedRoutingConditionContentPicker.propTypes = {
  data: PropTypes.object, // eslint-disable-line react/forbid-prop-types
  path: PropTypes.string.isRequired
};

export default props => (
  <AvailableRoutingQuestionsQuery pageId={props.pageId}>
    {innerProps => (
      <UnwrappedRoutingConditionContentPicker {...innerProps} {...props} />
    )}
  </AvailableRoutingQuestionsQuery>
);
