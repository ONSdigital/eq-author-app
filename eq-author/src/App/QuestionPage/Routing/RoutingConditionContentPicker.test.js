import React from "react";
import { shallow } from "enzyme";
import { UnwrappedRoutingConditionContentPicker } from "App/QuestionPage/Routing/RoutingConditionContentPicker";

const render = props =>
  shallow(<UnwrappedRoutingConditionContentPicker {...props} />);

describe("RoutingConditionContentPicker", () => {
  let wrapper, props;
  beforeEach(() => {
    props = {
      path: "foobar",
      onSubmit: jest.fn()
    };

    wrapper = render(props);
  });

  it("should render", () => {
    expect(wrapper).toMatchSnapshot();
  });
});
