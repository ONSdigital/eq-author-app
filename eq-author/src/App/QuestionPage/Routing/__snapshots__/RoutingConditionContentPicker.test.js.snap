// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoutingConditionContentPicker should render 1`] = `
<UnwrappedContentPickerSelect
  contentTypes={
    Array [
      "Question",
    ]
  }
  name="questionPageId"
  onSubmit={[MockFunction]}
  questionData={Array []}
  selectedContentDisplayName="Please select..."
/>
`;
